game/game/damagecalculator.cpp

```diff
@@ -51,17 +51,44 @@ DamageCalculator::Val DamageCalculator::damageFall(Npc& npc, float speed) {
   }
 
 DamageCalculator::Val DamageCalculator::rangeDamage(Npc& nsrc, Npc& nother, const Bullet& b, const CollideMask bMsk) {
-  bool invinsible = !checkDamageMask(nsrc,nother,&b);
-  if(b.pathLength() > float(MaxBowRange) * b.hitChance() && b.hitChance()<1.f)
-    return Val(0,false,invinsible);
+  float dist       = b.pathLength();
+  bool  noHit      = dist>float(MaxMagRange);
+  bool  invincible = !checkDamageMask(nsrc,nother,&b);
+  auto  dmg        = b.damage();
+
+  if(!b.isSpell()) {
+    auto& script    = nsrc.world().script();
+    int   hitChance = int(script.rand(100));
+    int   hitCh     = 0;
+    bool  g2        = Gothic::inst().version().game==2;
+    float critRange = g2 ? CritBowRange : (CritBowRange + 500);
+    float skill     = b.hitChanceVal();
+
+    if(dist<critRange)
+      hitCh = int((skill - 100.f) / critRange * dist + 100.f + 0.5f); else
+      hitCh = int(skill / (critRange - float(MaxBowRange)) * (dist - float(MaxBowRange)) + 0.5f);
+
+    noHit = (dist>float(MaxBowRange) || hitCh<=hitChance);
+
+    auto w = nsrc.inventory().activeWeapon();
```

Comment:
This is not reliable. `activeWeapon` can be anything, by the time bullet hits.
If you need to access `talentValue` - you can embed it into bullet, similar to hit chance